hera_cal/nucal.py

```diff
@@ -0,0 +1,464 @@
+
+    def add_radial_group(self, group):
+        """Adds a radially redundant group to the list of radially redundant groups stored in this object.
+        First checks to if the group is radially redundant, then adds the group to an existing group if a
+        group with the same heading already exists or appends the group if the heading is new.
+
+        Parameters:
+        ----------
+        group : list
+            List of baseline tuples to be added to the list of radially redundant groups
+        """
+        # Check to make sure the new group is radially redundant
+        self._check_new_group(group)
+            
+        # If group with same heading already exists, add it to that group. Otherwise, append the group to the list
+        for bl in self._mapped_spectral_reds:
+            if is_same_orientation(group[0], bl, self.antpos, blvec_error_tol=self.blvec_error_tol) and bl[-1] == group[0][-1]:
+                index = self._radial_groups.index(self._mapped_spectral_reds[bl])
+                self._radial_groups[index] += group
+                self._radial_groups[index] = list(set(self._radial_groups[index]))
+                break
+        else:
+            self._radial_groups.append(group)
+
+        # Add baseline lengths to length dictionary
+        for bl in group:
+            ant1, ant2, _ = bl
+            blmag = np.linalg.norm(self.antpos[ant2] - self.antpos[ant1])
+            self.baseline_lengths[bl] = blmag
+
+        # Reset the group now that radially redundant groups have changed
+        self._reset_mapping_dictionaries()
+
+    def __len__(self):
+        """Get number of frequency redundant groups"""
+        return len(self._radial_groups)
+
+    def __getitem__(self, index):
+        """Get RadialRedundantGroup object from list of unique orientations"""
+        return self._radial_groups[index]
+
+    def __setitem__(self, index, value):
+        """
+        Set value at index in _radial_groups. Also raises an error if a baseline with the same
+        heading is already in the list of radially redundant groups
+        """
+        # Check to make sure the new group is radially redundant
+        self._check_new_group(value)
+            
+        for bl in self._mapped_spectral_reds:
+            if is_same_orientation(value[0], bl, self.antpos, blvec_error_tol=self.blvec_error_tol) and bl[-1] == value[0][-1]:
+                raise ValueError('Radially redundant group with same orientation and polarization already exists in the data')
+                
+        # Add group at index
+        self._radial_groups[index] = value
+
+        # Add baseline lengths to length dictionary
+        for bl in value:
+            ant1, ant2, _ = bl
+            blmag = np.linalg.norm(self.antpos[ant2] - self.antpos[ant1])
+            self.baseline_lengths[bl] = blmag
+
+        # Reset baseline mapping to spectrally redundant groups
+        self._reset_mapping_dictionaries()
```

Comment:
Having to do this every time you append is very slow. Can you instead just update the mapping dictionaries?

Reply:
Good point. I just replace this with appending to dictionaries directly